src/recipe-selector/recipe-selector.ts

```diff
@@ -11,14 +11,54 @@
 import {Recipe} from '../runtime/recipe/recipe.js';
 import {Dictionary} from '../runtime/hot.js';
 
+// One entry in the lookup table, i.e. one trigger and the recipe it invokes.
+export class Match {
+  constructor(readonly trigger: [string, string][], readonly recipe: Recipe) {}
+
+  // The definition of a matching trigger is embodied in this function.
+  // This first implementation matches if every key-value pair in the trigger
+  // is present in the request. The request may contain other pairs, and the
+  // order of the matching pairs need not be the same as the order in the trigger. 
+  matches(request: [string, string][]): boolean {
+    if (this.trigger.length > request.length) return false;
+    for (const pair of this.trigger) {
+      const found = request.findIndex(r => r[0] === pair[0] && r[1] === pair[1]);
+      if (found === -1) return false;
+    }
+    return true;
+  }
+}
+
 export class RecipeSelector {
+    private _table: Match[] = [];
+
+    // For testing only
+    get table() {
+      return this._table;
+    }
+
     // Only recipes with “@trigger” annotations get included in the lookup table.
-    // Other recipes are silently ignored.
-    // Specific semantics of overlapping triggers are still TBD.
+    // Other recipes are silently ignored. This relies on the triggers array
+    // being always present but empty if there are no triggers.
+    // The order of the Matches in the lookup table preserves the order of the
+    // Recipes in the constructor argument, and within each recipe the order of
+    // the triggers in the triggers field of the recipe.
     constructor(readonly recipes: Recipe[]) {
+      recipes.forEach(recipe => {
+        recipe.triggers.forEach(trigger => {
+          if (this._table.find(match => match.matches(trigger))) {
+            throw new Error('Duplicate recipe trigger!');
```

Comment:
in the design doc you mentioned checking recipes for subset triggers, but matches works in only one direction, is this intentional?

Reply:
That changed, as it wasn't sufficient and too restrictive. The doc has been updated. Duplicates are disallowed, and the first match wins. This allows more specific triggers to come first, e.g. "app foo, action bar", with a more general catch-all later, e.g. "app foo".